### PR TITLE
TW-886: Configuration contact permission

### PR DESCRIPTION
**Pull request by nqhhdev**
### Issue:

- #886 

### Resolved

https://github.com/linagora/twake-on-matrix/assets/99852347/2c186b4a-007f-45dd-900a-b5525b4d1760

https://github.com/linagora/twake-on-matrix/assets/99852347/4af6f59e-f5c4-4298-b9b2-68a51643204d

https://github.com/linagora/twake-on-matrix/assets/99852347/014b72af-f48b-479e-a057-0a831568d6e8

https://github.com/linagora/twake-on-matrix/assets/99852347/5782f61a-3050-48f1-a178-8484b6315db0

#### WEB

https://github.com/linagora/twake-on-matrix/assets/99852347/fc8a212f-b4d1-4ccf-845d-ac76b9413898

### CODE REVIEW

File: lib/domain/contact_manager/contacts_manager.dart [hunk 1]
```diff
@@ -4,20 +4,39 @@ import 'package:fluffychat/config/app_config.dart';
...
 
+enum DisplayWaringContactsBannerState {
+  show,
+  hide,
+}
+
 class ContactsManager {
   SyncContactsState _syncContactsState = SyncContactsState.initial;
 
+  DisplayWaringContactsBannerState _displayWaringContactsBannerState =
```
Comment (drminh2807) [inline lib/domain/contact_manager/contacts_manager.dart:28]:
```suggestion
  DisplayWarningContactsBannerState _displayWarningContactsBannerState =
```

File: lib/pages/contacts_tab/contacts_tab_body_view.dart [hunk 2]
```diff
@@ -25,6 +26,16 @@ class ContactsTabBodyView extends StatelessWidget {
             color: Colors.black.withOpacity(0.15),
           ),
         ),
+        SliverToBoxAdapter(
+          child: ContactsWarningBannerView(
```
Comment (drminh2807) [inline lib/pages/contacts_tab/contacts_tab_body_view.dart:30]: duplicate code, ContactsWarningBannerView should have own controller and contacts manager

File: lib/domain/contact_manager/contacts_manager.dart [hunk 3]
```diff
@@ -4,20 +4,39 @@ import 'package:fluffychat/config/app_config.dart';
...
+      PlatformInfos.isMobile
+          ? DisplayWaringContactsBannerState.show
+          : DisplayWaringContactsBannerState.hide;
+
+  bool _isNotDisplayWarningAgainCacheMemory = false;
```
Comment (drminh2807) [inline lib/domain/contact_manager/contacts_manager.dart:33]: I think need only one variable
```
final showPermissionNotifier = ValueNotifier(PlatformInfos.isMobile)
```
other controllers also no need same value notifier

Comment (nqhhdev) [reply lib/domain/contact_manager/contacts_manager.dart:33]: What? why need to care UI in here?

File: lib/domain/contact_manager/contacts_manager.dart [hunk 4]
```diff
@@ -4,20 +4,39 @@ import 'package:fluffychat/config/app_config.dart';
...
 enum SyncContactsState {
   initial,
   synchronizing,
   synchronized,
 }
 
+enum DisplayWaringContactsBannerState {
+  show,
+  hide,
+}
```
Comment (hoangdat) [inline lib/domain/contact_manager/contacts_manager.dart:23]: move to `ContactViewManagerMixin`

File: lib/domain/contact_manager/contacts_manager.dart (hunk 1 above)
Comment (hoangdat) [reply lib/domain/contact_manager/contacts_manager.dart:28]: idem

File: lib/domain/contact_manager/contacts_manager.dart (hunk 3 above)
Comment (hoangdat) [reply lib/domain/contact_manager/contacts_manager.dart:33]: idem

File: lib/presentation/mixins/contacts_controller_mixin.dart
Comment (hoangdat) [inline lib/presentation/mixins/contacts_controller_mixin.dart:1]: Please change the name to `ContactsViewControllerMixin`

File: lib/domain/model/contact/contacts_manager_agruments.dart [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import 'package:equatable/equatable.dart';
+import 'package:fluffychat/domain/contact_manager/contacts_manager.dart';
+import 'package:fluffychat/domain/model/contact/contact.dart';
+
+class ContactsManagerArguments extends Equatable {
```
Comment (hoangdat) [inline lib/domain/model/contact/contacts_manager_agruments.dart:5]: Create an abstraction first, in the future a lot of kinds of things need to add to Contact Manager

File: lib/presentation/mixins/contacts_controller_mixin.dart [hunk 6]
```diff
@@ -77,11 +75,34 @@ mixin class ContactsControllerMixin {
...
+        arg.displayWaringContactsBannerState ==
+            DisplayWaringContactsBannerState.show;
+  }
+
+  void closeContactsWarningBanner() {
+    contactsManager.closeContactsWarningBanner();
+  }
+
+  void _listenContactsDataChange() {
+    contactsManager.contactsStream.stream.listen((arg) {
```
Comment (hoangdat) [inline lib/presentation/mixins/contacts_controller_mixin.dart:94]: should have a `sub` to dispose when controller dispose

File: lib/domain/contact_manager/contacts_manager.dart [hunk 7]
```diff
@@ -27,22 +46,43 @@ class ContactsManager {
...
+  void initialSynchronizeContacts() async {
+    if (PlatformInfos.isMobile && !_isNotDisplayWarningAgainCacheMemory) {
+      await _handleRequestContactsPermission();
+    }
     if (!firstSynchronized) {
       _syncContactsState = SyncContactsState.synchronizing;
       _getAllContacts();
+    } else {
+      _handleContactsHasDataChange();
+    }
```
Comment (hoangdat) [inline lib/domain/contact_manager/contacts_manager.dart:67]: use the switch case more readable

File: lib/domain/contact_manager/contacts_manager.dart [hunk 8]
```diff
@@ -4,20 +4,39 @@ import 'package:fluffychat/config/app_config.dart';
...
+          : DisplayWaringContactsBannerState.hide;
+
+  bool _isNotDisplayWarningAgainCacheMemory = false;
+
   final _getAllContactsInteractor = getIt.get<GetAllContactsInteractor>();
 
+  final PermissionHandlerService permissionHandlerService =
+      PermissionHandlerService();
+
   Set<Contact> _tomContacts = {};
```
Comment (hoangdat) [inline lib/domain/contact_manager/contacts_manager.dart:40]: why it is not `ValueNotifier`? It makes app freeze a lot. Update it whenever we complete it. Avoid

**Review by drminh2807 — APPROVED** [review]

**Review by sherlockvn — APPROVED** [review]

File: lib/pages/new_group/widget/contacts_selection_list.dart [hunk 9]
```diff
@@ -12,6 +14,7 @@ class ContactsSelectionList extends StatelessWidget {
   final Function() onSelectedContact;
   final List<String> disabledContactIds;
   final TextEditingController textEditingController;
+  final ContactsManager contactsManager;
```
Comment (hoangdat) [inline lib/pages/new_group/widget/contacts_selection_list.dart:17]: it should be better for singleton instance
```suggestion
  final ContactsManager contactsManager = getIt<ContactsManager>();
```

File: lib/pages/new_private_chat/widget/expansion_list.dart [hunk 10]
```diff
@@ -17,6 +20,8 @@ class ExpansionList extends StatelessWidget {
   final Function(BuildContext context, PresentationContact contact)
       onContactTap;
   final TextEditingController textEditingController;
+  final ValueNotifier<bool> isShowContactsWarningBannerNotifier;
+  final ContactsManager contactsManager;
```
Comment (hoangdat) [inline lib/pages/new_private_chat/widget/expansion_list.dart:24]: idem

File: lib/domain/contact_manager/contacts_manager.dart [hunk 11]
```diff
@@ -26,27 +39,70 @@ class ContactsManager {
...
+  bool _isExternalContact(String keyword) =>
+      keyword.isValidMatrixId && keyword.startsWith("@");
+
+  StreamController<Either<Failure, Success>> contactsStream =
+      StreamController<Either<Failure, Success>>.broadcast();
 
-  bool get firstSynchronized =>
-      _syncContactsState == SyncContactsState.synchronized;
+  StreamController<Either<Failure, Success>> waringBannerStream =
+      StreamController<Either<Failure, Success>>.broadcast();
```
Comment (hoangdat) [inline lib/domain/contact_manager/contacts_manager.dart:53]: why not use one `stream`?

File: lib/pages/new_group/widget/contacts_selection_list.dart (hunk 9 above)
Comment (nqhhdev) [reply lib/pages/new_group/widget/contacts_selection_list.dart:23]: No need, I'll remove it

File: lib/pages/new_private_chat/widget/expansion_list.dart (hunk 10 above)
Comment (nqhhdev) [reply lib/pages/new_private_chat/widget/expansion_list.dart:30]: remove it

File: lib/domain/contact_manager/contacts_manager.dart [hunk 12]
```diff
@@ -63,4 +106,59 @@ class ContactsManager {
...
+  void closeContactsWarningBanner() {
+    _doNotShowWarningContactsBannerAgain = true;
+    warningBannerStateStream.add(WarningContactsBannerState.notDisplayAgain);
+  }
+
+  void searchContacts(String keyword) {
+    final contactsMatched = tomContacts.toList().searchContacts(keyword);
+    if (contactsMatched.isEmpty && _isExternalContact(keyword)) {
+      contactsStream.add(
+        Right(SearchExternalContactsSuccessState(keyword: keyword)),
```
Comment (hoangdat) [inline lib/domain/contact_manager/contacts_manager.dart:119]: why not have an exactly empty state?

File: lib/domain/contact_manager/contacts_manager.dart [hunk 13]
```diff
@@ -26,25 +39,55 @@ class ContactsManager {
...
+
+  void initialSynchronizeContacts() async {
+    if (PlatformInfos.isMobile &&
+        !_doNotShowWarningContactsBannerAgain &&
+        _firstSynchronizing) {
+      contactsStream.add(const Right(ContactsLoading()));
+      await _handleRequestContactsPermission();
+    }
+    switch (_syncContactsState) {
+      case SyncContactsState.synchronizing:
```
Comment (hoangdat) [inline lib/domain/contact_manager/contacts_manager.dart:63]: I dont think we can do the duplicated tasks if it is in `synchronizing`

Comment (hoangdat) [reply lib/domain/contact_manager/contacts_manager.dart:63]: my fault. Done

File: lib/domain/contact_manager/contacts_manager.dart [hunk 14]
```diff
@@ -26,25 +39,55 @@ class ContactsManager {
...
+      }
     }
   }
 
   void _getAllContacts() {
-    contactsStreamSubscription = _getAllContactsInteractor
+    _getAllContactsInteractor
         .execute(limit: AppConfig.maxFetchContacts)
         .listen((event) {
+      contactsStream.add(event);
```
Comment (hoangdat) [inline lib/domain/contact_manager/contacts_manager.dart:90]: please update the list first, then broadcast the thing later

**Review by hoangdat — APPROVED** [review]

**Pull request merged**